### PR TITLE
✨ feat: implements update and delete for virtual devices

### PR DESCRIPTION
**Pull request by pallabpain**
### Description
This pull request improves the support for onboarding HWIL devices via device manifests. We implement the `update_object` and `delete_object` such that HWIL devices are re-onboarded or deleted via the `rio apply` and `rio delete` commands.

Take an example where the rapyuta.io device is present but the corresponding HWIL device pod was evicted, we will have to re-onboard it. On re-apply the manifest, the `update_object` handler will take care of running the onboarding script against on the said device. If in case the device was deleted as a part of some cleanup, then re-applying will ensure that the device is re-created and onboarded.

Similarly, `rio delete device.yaml` will ensure that both rapyuta.io and the HWIL device are cleaned up.

### Testing
Manifest used for testing
```yaml
apiVersion: "apiextensions.rapyuta.io/v1"
kind: "Device"
metadata:
  name: "test01"
  labels:
    purpose: testing
    app: test
spec:
  python: "3"
  rosDistro: "melodic"
  virtual:
    enabled: True
    product: "sootballs"
    arch: "amd64"
    os: "ubuntu"
    codename: "focal"
    highperf: False
  docker:
    enabled: True # Required
    rosbagMountPath: "/opt/rapyuta/volumes/rosbag"
```

### CODE REVIEW

File: riocli/device/util.py [hunk 1]
```diff
@@ -194,13 +194,44 @@ def create_hwil_device(spec: dict, metadata: dict) -> Munch:
...
+            break
+
+    if device_id is None:
+        raise DeviceNotFound(message='hwil device not found')
+
+    client = new_hwil_client()
+    try:
+        client.delete_device(device_id)
+    except Exception as e:
+        raise e
```
Comment (ankitrgadiya) [inline riocli/device/util.py:231]: suggestion: This is equivalent to not handling exception.

File: riocli/device/model.py [hunk 2]
```diff
@@ -48,33 +44,84 @@ def find_object(self, client: Client) -> bool:
...
+        if device['status'] in ('ONLINE', 'INITIALIZING'):
+            return
+
+        device_labels = device.get('labels', {})
+        # Convert list to dict for easy access.
+        device_labels = {l['key']: l for l in device_labels}
+
+        # Otherwise, re-onboard the device.
         try:
+            hwil_response = create_hwil_device(self.spec, self.metadata)
```
Comment (ankitrgadiya) [inline riocli/device/model.py:92]: suggestion: Consider checking if the device from the metadata already exists. In the creation, the failure might have happened after the device was created (maybe during the onboarding).

File: riocli/device/model.py [hunk 3]
```diff
@@ -48,33 +44,84 @@ def find_object(self, client: Client) -> bool:
...
+                device_uuid = find_device_guid(client, self.metadata.name)
+                device = client.get_device(device_uuid)
+                delete_hwil_device(device)
+            except Exception as e:
+                raise e
+
+        try:
+            obj.delete()
+        except Exception as e:
+            raise e
```
Comment (ankitrgadiya) [inline riocli/device/model.py:124]: suggestion: These are equivalent to not handling exception. Consider simplifying.

**Review by ankitrgadiya — CHANGES_REQUESTED** [review]

File: riocli/device/model.py (hunk 3 above)
Comment (pallabpain) [reply riocli/device/model.py:124]: Since we are handling the exceptions in the base methods, I am removing the try-except blocks from this file.

**Review by pallabpain — COMMENTED** [review]
@ankitrgadiya @amitsingh21 I've addressed the comments. Used this manifest for testing
```yaml
apiVersion: "apiextensions.rapyuta.io/v1"
kind: "Device"
metadata:
  name: "test01"
  labels:
    purpose: testing
    app: test
spec:
  python: "3"
  rosDistro: "melodic"
  virtual:
    enabled: True
    product: "sootballs"
    arch: "amd64"
    os: "ubuntu"
    codename: "focal"
    highperf: False
  docker:
    enabled: True # Required
    rosbagMountPath: "/opt/rapyuta/volumes/rosbag"
```

**Review by ankitrgadiya — APPROVED** [review]
Looks good, you can squash the commits.

**Pull request merged**